### PR TITLE
Add Goodbye Message When Disconnecting With Peers

### LINKED ISSUE #4814: Nodes Should Send Goodbye Messages When Shutting Down
Currently when we shut down our node, we just disconnect with all our peers at once. The proper way would be to send a goodbye message to them and then disconnect with them

### PR DESCRIPTION
**Pull request by nisdas**
Resolves #4814 

- [x] We now send goodbye messages to clients, when we disconnect to them in case we hit our peer limit. 
- [x] Also we do not add timeout's to our primary context when passing it down to its sub-methods. This leads to the child methods adopting the parent context's deadlines and leads to early timeouts when performing handshakes with other peers. 

cc @q9f , @AgeManning

### CODE REVIEW

File: beacon-chain/p2p/handshake.go [hunk 1]
```diff
@@ -28,10 +28,15 @@ func (s *Service) AddConnectionHandler(reqFunc func(ctx context.Context, id peer
 			}
 			s.peers.Add(nil /* ENR */, conn.RemotePeer(), conn.RemoteMultiaddr(), conn.Stat().Direction)
 			if len(s.peers.Active()) >= int(s.cfg.MaxPeers) {
-				log.WithField("reason", "at peer limit").Trace("Ignoring connection request")
-				if err := s.Disconnect(conn.RemotePeer()); err != nil {
-					log.WithError(err).Error("Unable to disconnect from peer")
-				}
+				go func() {
+					log.WithField("reason", "at peer limit").Trace("Ignoring connection request")
+					if err := goodbyeFunc(context.Background(), 2, conn.RemotePeer()); err != nil {
```
Comment (shayzluf) [inline beacon-chain/p2p/handshake.go:33]: can we please name "2" what is this code?

**Comment by protolambda** [comment]
My topaz prysm testnet node (on debug log level) has logged hundreds of thousands of lines over the past days, most of which are level=debug msg="Sending Goodbye message to peer" Reason="fault/error" peer=16Uiu2HAm4TsWtLtCmsS4szLRUacQiNFVAcJxZPHjPiDS9XnHLRC6 prefix=sync (with different peer IDs).

Also getting a lot "Failed to decode goodbye stream message" error="stream reset" peer=16Uiu2HAm7XYQgKi27yjvd1G433rJJzpCyrKvzbjeE54DA4n8y6FP prefix=sync topic="/eth2/beacon_chain/req/goodbye/1/ssz_snappy"

Also, the "failed to decode goodbye" sometimes logs up to 8 times, the exact same

Logs excerpt (literally hundreds of thousands of lines of logs, can't share it all, but you get the idea): https://gist.github.com/protolambda/a087de4f44a03deeff0c2a78f50c07ca

Can it be that this functionality is horribly broken in some unexpected way?

**Comment by protolambda** [comment]
4 days of running in debug mode:
```
$ sudo docker logs 8cb989bbd69b 2>&1 | cut -c34- | grep "Goodbye" | wc -l
780398
$ sudo docker logs 8cb989bbd69b 2>&1 | cut -c34- | grep "Sending Goodbye message to peer\" Reason=\"fault/error" | wc -l
780509
$ sudo docker logs 8cb989bbd69b 2>&1 | cut -c34- | grep "Goodbye" | wc -l
780599
$ sudo docker logs 8cb989bbd69b 2>&1 | cut -c34- | wc -l
1889091
```
That's a lot of log messages...

**Comment by prestonvanloon** [comment]
Are you at max peers? You'll kick anyone that tries to connect to you if you have max peers and they are probably retrying anyway.

**Comment by prestonvanloon** [comment]
Screenshot from 2020-05-09 14-36-00

Seeing this often in prod too. @nisdas It may be logging too much or sending multiple good bye to the same peer.

**Comment by protolambda** [comment]
At max peers yes, around 100 (changed upwards from default). docker image 9a7fbadca570, commit 2eac24cb79b0c847a74e3f8395ab5808c4059dce

**Comment by protolambda** [comment]
Still, if that's max-peers, and it logs goodbye because of failed connect, then that's almost 800.000 connection attempts to my node in 4 days. So maybe the goodbye code is correct, but the connection attempts are way too much.

**Comment by prestonvanloon** [comment]
Confirmed we are logging more than once per peer. Possibly sending multiple goodbye messages. Some peers logging as much as 50+ on a single node.

Screenshot from 2020-05-09 14-42-36


Graph link for @nisdas https://kibana.prylabs.network/goto/873081efd6a6753438a9e8f253e6fd1a

**Comment by protolambda** [comment]
Even at 50 per peer, that's 16000 connection attempts

**Pull request merged**